<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Document</title>
</head>
<body>
  <!-- <script src="./basePromise.js"></script>
  <script>
    var promise = new basePromise((resolve, reject) => {
      resolve('成功');
    }).then(
      (data) => {
        console.log('success', data)
      },
      (err) => {
        console.log('faild', err)
      }
    )
  </script> -->
  <!-- 
    以上是实现了基础版本的手写Promise，控制台输出’success 成功‘
    但是不支持异步操作，以下代码不会有任何输出
    因为 promise 调用 then 方法时，当前的 promise 并没有成功，一直处于 pending 状态。
    所以如果当调用 then 方法时，当前状态是 pending，我们需要先将成功和失败的回调分别存放起来，
    在executor()的异步任务被执行时，触发 resolve 或 reject，依次调用成功或失败的回调。
  -->
  <!-- <script>
    var promise = new basePromise((resolve, reject) => {
      // 传入一个异步操作
      setTimeout(() => {
        resolve('成功');
      },1000);
    }).then(
      (data) => {
        console.log('success', data)
      },
      (err) => {
        console.log('faild', err)
      }
    )
  </script> -->

  <!-- 结合以上分析，我们再优化一下代码 -->
  <script src="./promise.js"></script>
  <!-- 再测试一下 -->
  <script>
    var promise = new myPromise((resolve, reject) => {
      // 传入一个异步操作
      setTimeout(() => {
        resolve('成功');
      },1000);
    }).then(
      (data) => {
        console.log('success', data) // 一秒后输出 'success 成功'
      },
      (err) => {
        console.log('faild', err)
      }
    )
  </script>
</body>
</html>